### PR TITLE
Add support for hanging nodes to the NedelecSZ class.

### PR DESCRIPTION
**Pull request by kinnewig**
Back on the deal.II user workshop in Hanover, I promised to create a pull request to add the hanging node support for FE_NedelecSZ elements soon. However, I was busy working on the Tpetra wrappers instead, as I needed the Tpetra interface for another project. But now I finally found the time to clean up the code and make some final changes.

I am happy to create this pull request, introducing support for hanging nodes to the FE_NedelecSZ elements!

The theory behind the implementation is covered in my paper: [Algorithmic realization of the solution to the sign conflict problem for hanging nodes on hp-hexahedral Nédélec elements](https://arxiv.org/abs/2306.01416).

I have not added any tests to this pull request jet since the tests presented in the paper take too long to evaluate (especially in 3D since we need a polynomial degree of p = 2 to cover every case). So, if you have any suggestions for more performant tests, please let me know, and I will be happy to include them!

### CODE REVIEW

File: source/dofs/dof_tools_constraints.cc [hunk 1]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+              AssertDimension(cell->face(face)->n_active_fe_indices(), 1);
+
+              Assert(cell->face(face)->fe_index_is_active(
+                       cell->active_fe_index()) == true,
+                     ExcInternalError());
+
+              for (unsigned int c = 0; c < cell->face(face)->n_children(); ++c)
+                if (!cell->neighbor_child_on_subface(face, c)->is_artificial())
+                  AssertDimension(
+                    cell->face(face)->child(c)->n_active_fe_indices(), 1);
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1294]: Maybe, just combine with the loop below?

File: source/dofs/dof_tools_constraints.cc [hunk 2]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+
+              // get the polynomial degree
+              unsigned int degree(fe.degree);
+
+              // get the number of DoFs on mother and children;
+              // number of DoFs on the mother
+              const unsigned int n_dofs_on_mother = fe.n_dofs_per_face(face);
+              dofs_on_mother.resize(n_dofs_on_mother);
+
+              // number of intern lines of the children;
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1349]:
```suggestion
              // number of internal lines of the children;
```

File: source/dofs/dof_tools_constraints.cc [hunk 3]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+              // for more details see: Description of the class
+              // GeometryInfo<dim>
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1351]:
```suggestion
              // for more details see description of the
              // GeometryInfo<dim> class
```

File: source/dofs/dof_tools_constraints.cc [hunk 4]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+              // .................
+              // .       |       .
+              // .  c2   1   c3  .
+              // .       |       .
+              // .---2---+---3---.
+              // .       |       .
+              // .  c0   0   c1  .
+              // .       |       .
+              // .................
+              const unsigned int n_intern_lines_on_children = 4;
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1361]:
```suggestion
              const unsigned int n_internal_lines_on_children = 4;
```

File: source/dofs/dof_tools_constraints.cc [hunk 5]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+              // do not resize(), but reserve() and use push_back later.
+              dofs_on_children.clear();
+              dofs_on_children.reserve(n_dofs_on_children);
+
+              Assert(n_dofs_on_mother == fe.constraints().n(),
+                     ExcDimensionMismatch(n_dofs_on_mother,
+                                          fe.constraints().n()));
+              Assert(n_dofs_on_children == fe.constraints().m(),
+                     ExcDimensionMismatch(n_dofs_on_children,
+                                          fe.constraints().m()));
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1398]:
```suggestion
              AssertDimension(n_dofs_on_mother, fe.constraints().n());
              AssertDimension(n_dofs_on_children, fe.constraints().m());
```

File: source/dofs/dof_tools_constraints.cc [hunk 6]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+              // fill the DoF on the children:
+              Assert(dof_handler.get_triangulation()
+                         .get_anisotropic_refinement_flag() ||
+                       ((this_face->child(0)->vertex_index(3) ==
+                         this_face->child(1)->vertex_index(2)) &&
+                        (this_face->child(0)->vertex_index(3) ==
+                         this_face->child(2)->vertex_index(1)) &&
+                        (this_face->child(0)->vertex_index(3) ==
+                         this_face->child(3)->vertex_index(0))),
+                     ExcInternalError());
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1435]: What's going on here? Why is it OK if there has been some anisotropic refinement?

File: source/dofs/dof_tools_constraints.cc [hunk 7]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+                  this_face->child(0)->line(3)->dof_index(dof, fe_index));
+
+              for (unsigned int dof = 0; dof < fe.n_dofs_per_line(); ++dof)
+                dofs_on_children.push_back(
+                  this_face->child(1)->line(3)->dof_index(dof, fe_index));
+
+              // DoFs on the bordering lines:
+              // DoFs on the exterior lines to the children; the order of
+              // these lines is shown above (see n_external_lines_on_children)
+              for (unsigned int line = 0; line < 4; ++line)
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1462]:
```suggestion
              for (unsigned int line = 0; line < /* lines per face */ 4; ++line)
```
or
```suggestion
              for (unsigned int line = 0; line < GeometryInfo<dim>::lines_per_face; ++line)
```

File: source/dofs/dof_tools_constraints.cc [hunk 8]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+
+              // DoFs on the faces of the four children:
+              for (unsigned int child = 0; child < n_of_children; ++child)
+                {
+                  // skip artificial cells
+                  if (cell->neighbor_child_on_subface(face, child)
+                        ->is_artificial())
+                    continue;
+
+                  for (unsigned int dof = 0; dof != fe.n_dofs_per_quad(face);
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1477]:
```suggestion
                  for (unsigned int dof = 0; dof < fe.n_dofs_per_quad(face);
```

File: source/dofs/dof_tools_constraints.cc [hunk 9]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+              for (unsigned int dof = 0; dof < fe.n_dofs_per_line(); ++dof)
+                dofs_on_children.push_back(
+                  this_face->child(1)->line(3)->dof_index(dof, fe_index));
+
+              // DoFs on the bordering lines:
+              // DoFs on the exterior lines to the children; the order of
+              // these lines is shown above (see n_external_lines_on_children)
+              for (unsigned int line = 0; line < 4; ++line)
+                for (unsigned int child = 0; child < 2; ++child)
+                  for (unsigned int dof = 0; dof != fe.n_dofs_per_line(); ++dof)
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1464]:
```suggestion
                  for (unsigned int dof = 0; dof < fe.n_dofs_per_line(); ++dof)
```
Is there a reason for using a different style here?

File: source/dofs/dof_tools_constraints.cc [hunk 10]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+                    dofs_on_children.push_back(
+                      this_face->child(child)->dof_index(dof, fe_index));
+                } // rof: child
+
+              // consistency check:
+              // note: we can get fewer DoFs when we have artificial cells
+              Assert(dofs_on_children.size() <= n_dofs_on_children,
+                     ExcInternalError());
+
+              // As the Nedelec elements are oriented we need to take care of
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1488]:
```suggestion
              // As the Nedelec elements are oriented, we need to take care of
```

File: source/dofs/dof_tools_constraints.cc [hunk 11]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+              // the orientation of the lines.
+              // Remark: "False" indicates the line is not flipped.
+              //         "True" indicates the line is flipped.
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1491]:
```suggestion
              // Remark: "false" indicates the line is not flipped.
              //         "true" indicates the line is flipped.
```

File: source/dofs/dof_tools_constraints.cc [hunk 12]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+                      (line_mother * 2) * fe.n_dofs_per_line();
+                    for (unsigned int row = 0; row < fe.n_dofs_per_line();
+                         ++row)
+                      for (unsigned int i = 0; i < 4 * fe.n_dofs_per_line();
+                           ++i)
+                        constraints_matrix[row + row_start][i] =
+                          fe.constraints()(row + row_mother, i);
+                  }
+
+                for (unsigned int line = 0; line < 4; ++line)
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1645]: internal lines?

File: source/dofs/dof_tools_constraints.cc [hunk 13]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+                             i < dofs_on_mother.size();
+                             ++i)
+                          constraints_matrix[row + row_start][i] =
+                            fe.constraints()(row + row_mother, i);
+                      }
+                  }
+
+                // copy the weights for exterior edges
+                unsigned int row_offset = 4 * fe.n_dofs_per_line();
+                for (unsigned int line = 0; line < 8; line++)
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1664]: external lines?

File: source/dofs/dof_tools_constraints.cc [hunk 14]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+                    for (unsigned int row = row_offset;
+                         row < row_offset + fe.n_dofs_per_line();
+                         ++row)
+                      for (unsigned int i = 0; i < dofs_on_mother.size(); ++i)
+                        constraints_matrix[row + row_start][i] =
+                          fe.constraints()(row + row_mother, i);
+                  }
+
+                // copy the weights for the faces
+                row_offset = 12 * fe.n_dofs_per_line();
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1679]: total lines?

File: source/dofs/dof_tools_constraints.cc [hunk 15]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+                for (unsigned int face = 0; face < 4; ++face)
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1680]: faces per cell?

File: source/dofs/dof_tools_constraints.cc [hunk 16]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+                              direction_mother[line_j])
+                            {
+                              if ((tmp_i + tmp_j) % 2 == 0)
+                                { // symmetric
+                                  constraints_matrix[i][j] *= 1.0;
+                                }
+                              else
+                                { // anti-symmetric
+                                  constraints_matrix[i][j] *= -1.0;
+                                }
```
Comment (masterleinad) [inline source/dofs/dof_tools_constraints.cc:1720]:
```suggestion
                              if ((tmp_i + tmp_j) % 2 == 1)
                                { // anti-symmetric
                                  constraints_matrix[i][j] *= -1.0;
                                }
```
and elsewhere

**Review by masterleinad — COMMENTED** [review]
Just a more high-level review. We absolutely need tests to make sure all the nitty-gritty details are correct. 🙂 
In general, it would be good to avoid more of the magic numbers.

File: source/dofs/dof_tools_constraints.cc [hunk 17]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+        {
+          // artificial cells can at best neighbor ghost cells, but we're not
+          // interested in these interfaces
+          if (cell->is_artificial())
+            continue;
+
+          // loop over all faces:
+          for (const unsigned int f : cell->face_indices())
+            {
+              // check of the neighbor is refined; if so, we need to
```
[the remark this answers is not in the document]
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc]: Fixed that typo.

File: source/dofs/dof_tools_constraints.cc [hunk 18]
```diff
@@ -1070,6 +1101,1316 @@ namespace DoFTools
...
+                  Assert(cell->face(f)->child(c)->n_active_fe_indices() == 1,
+                         ExcInternalError());
+
+              // right now, all that is implemented is the case that both
+              // sides use the same FE
+              for (unsigned int c = 0; c < cell->face(f)->n_children(); ++c)
+                if (!cell->neighbor_child_on_subface(f, c)->is_artificial())
+                  Assert(cell->face(f)->child(c)->fe_index_is_active(
+                           cell->active_fe_index()) == true,
+                         ExcNotImplemented());
```
[the remark this answers is not in the document]
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc]: Yes, I combined those two loops into one. Furthermore, I added a #ifdef DEBUG (...) #endif, so those loops are only executed in debug mode.

File: source/dofs/dof_tools_constraints.cc (hunk 1 above)
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc]: Similar to above, I combined the loops and added #ifdef DEBUG (...) #endif.

File: source/dofs/dof_tools_constraints.cc (hunk 2 above)
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc]: Fixed that typo

File: source/dofs/dof_tools_constraints.cc (hunk 3 above)
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc]: Done!

File: source/dofs/dof_tools_constraints.cc (hunk 4 above)
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc]: I refactored that variable name.

File: source/dofs/dof_tools_constraints.cc (hunk 6 above)
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc]: Well, this is a copy-and-paste mistake. I added more asserts directly before opening this pull request. I intend to check that we have only isotropic refinement, but I forgot to adapt that assertion accordingly.
I have fixed that.

File: source/dofs/dof_tools_constraints.cc (hunk 7 above)
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc]: I tried to reduce the amount of magic numbers; therefore, I included your second suggestion.

File: source/dofs/dof_tools_constraints.cc (hunk 13 above)
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc]: Yes, and similar to the above, I added a comment to clarify that. Furthermore, I replaced the "4" with n_external_lines_on_children to reduce the amount of magic numbers.

**Comment by kinnewig** [comment]
@masterleinad, thank you so much for catching all the corrections! The code was developed incrementally over a long period, allowing several spelling mistakes to slip through.

I reduced the number of magic numbers by a fair amount. That should improve the readability of the code.

Also, I have developed some ideas for tests. I will add them as soon as I find the time to implement them.

**Comment by kinnewig** [comment]
I added a first test. The test is already very comprehensive and should cover most parts of the hanging node implementation.

It is a convergence test where we solve a real-valued Maxwell equation on a cylinder (or a disk in 2D). Moreover, we choose the boundary values such that we know the exact solution so that we can compare the numerical result with the exact solution.
We start on a mesh without any hanging nodes and add hanging nodes in the later iterations step. While doing this, we check that the numerical solution is close to the exact solution, and we check that when we add more and more cells, we converge against the exact solution.

The grid and the refinement are designed to cover most configurations, but we need to do 3 iteration steps. We could reduce it into two steps, but then one special case would not be covered. I am considering turning that special case into its own test case, anyhow.

This is already a very high-level test and, therefore, should catch many problems. The drawback is that the test has quite a high run time (just below 2 minutes on my system).

ConvergenceTest

File: source/fe/fe_nedelec_sz.cc [hunk 19]
```diff
@@ -1713,7 +1822,7 @@ FE_NedelecSZ<dim, spacedim>::fill_edge_values(
                             }
 
                           // Higher order edge shape function
-                          for (unsigned int i = 0; i < degree; ++i)
+                          for (unsigned int i = 1; i < degree; ++i)
```
[the remark this answers is not in the document]
Comment (kinnewig) [reply source/fe/fe_nedelec_sz.cc]: I removed this one; the original implementation was correct; I was just confused about something.

**Comment by kinnewig** [comment]
I turned all remarks from @agrayver into test cases. With that, this is covered comprehensively with test cases.

I would be happy if one of the core developers could review the code (maybe @kronbichler or @tamiko?).

At the moment, there is still one TODO left open. Currently, the implementation assumes four (or fewer) cells are adjacent to each line (in 3D). But I will fix that in a separate pull request.

File: source/dofs/dof_tools_constraints.cc [hunk 20]
```diff
@@ -1070,6 +1101,1290 @@ namespace DoFTools
...
+              for (unsigned int row = 0; row < n_dofs; ++row)
+                {
+                  for (unsigned int dof_i_on_mother = 0;
+                       dof_i_on_mother < n_dofs;
+                       ++dof_i_on_mother)
+                    {
+                      // We need to keep in mind that, if we use a FE_System
+                      // with multiple FE_NedelcSZ blocks inside, we need
+                      // to consider, that n_dofs depends on the number
+                      // of FE_NedelcSZ blocks used.
```
Comment (kronbichler) [inline source/dofs/dof_tools_constraints.cc:1218]:
```suggestion
                      // of FE_NedelecSZ blocks used.
```
I think there are a few more appearances of this kind, please fix them all.

File: source/dofs/dof_tools_constraints.cc [hunk 21]
```diff
@@ -1070,6 +1101,1290 @@ namespace DoFTools
...
+
+              // As the Nedelec elements are oriented, we need to take care of
+              // the orientation of the lines.
+              // Remark: "false" indicates the line is not flipped.
+              //         "true" indicates the line is flipped.
+
+              // Orientation - Lines:
+              // get the orientation from the edges from the mother cell
+              std::vector<bool> direction_mother(
+                GeometryInfo<dim>::lines_per_face, false);
```
Comment (kronbichler) [inline source/dofs/dof_tools_constraints.cc:1501]: I think you used `int` with value 1 or -1 for the 2D case above for `direction_mother`. Can you comment on the different choices, or unify to a similar data type in case they have the same underlying switches?

File: source/dofs/dof_tools_constraints.cc [hunk 22]
```diff
@@ -1070,6 +1101,1290 @@ namespace DoFTools
...
+              bool              mother_flip_y  = false;
+              bool              mother_flip_xy = false;
+              std::vector<bool> child_flip_x(n_children_per_face, false);
+              std::vector<bool> child_flip_y(n_children_per_face, false);
+              std::vector<bool> child_flip_xy(n_children_per_face, false);
+              const unsigned int
+                vertices_adjacent_on_face[GeometryInfo<dim>::vertices_per_face]
+                                         [2] = {{1, 2}, {0, 3}, {3, 0}, {2, 1}};
+
+              { // Mother
```
Comment (kronbichler) [inline source/dofs/dof_tools_constraints.cc:1565]: I think we typically do not place comments in the same line as code, unless the comment is short enough for a single line (whereas you here have 2 lines, albeit they are referring to different things).
```suggestion
              {
                // Mother
```

File: source/dofs/dof_tools_constraints.cc [hunk 23]
```diff
@@ -1070,6 +1101,1290 @@ namespace DoFTools
...
+                  child_flip_xy[child] = mother_flip_xy;
+                }
+
+              // copy the constraint matrix, since we need to modify that matrix
+              std::vector<std::vector<double>> constraints_matrix(
+                n_lines_on_children * fe.n_dofs_per_line() +
+                  n_children_per_face * fe.n_dofs_per_quad(),
+                std::vector<double>(dofs_on_mother.size(), 0));
+
+              { // copy the constraint matrix
```
Comment (kronbichler) [inline source/dofs/dof_tools_constraints.cc:1634]:
```suggestion
              {
```

File: source/fe/fe_nedelec_sz.cc [hunk 24]
```diff
@@ -45,6 +46,104 @@ FE_NedelecSZ<dim, spacedim>::FE_NedelecSZ(const unsigned int order)
 
   // Generate the 1-D polynomial basis.
   create_polynomials(order);
+
+  // Compute the face embedding.
+  FullMatrix<double> face_embeddings[GeometryInfo<dim>::max_children_per_face];
+
+  // TODO: The implementation assumes that all faces have the same
+  //       number of DoFs.
```
Comment (kronbichler) [inline source/fe/fe_nedelec_sz.cc:54]: Why did you set a `TODO` here? I think there are several more things to be done if we really wanted to use the NedelecSZ elements on something like a wedge or a pyramid, so I feel that we should say
```suggestion
  // The implementation assumes that all faces have the same
  // number of DoFs.
```

File: tests/fe/fe_nedelec_sz_hanging_nodes_convergence.cc [hunk 25]
```diff
@@ -0,0 +1,448 @@
...
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+// By Sebastian Kinnewig
```
Comment (kronbichler) [inline tests/fe/fe_nedelec_sz_hanging_nodes_convergence.cc:15]: Can you please remove this line? We avoid tagging files with particular authors as we see the library as owned and maintained by a collective of all deal.II authors, including you. Your contribution will obviously still be visible via the commit author/git annotation.

File: tests/fe/fe_nedelec_sz_hanging_nodes_convergence.cc [hunk 26]
```diff
@@ -0,0 +1,448 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2020 by the deal.II authors
```
Comment (kronbichler) [inline tests/fe/fe_nedelec_sz_hanging_nodes_convergence.cc:3]:
```suggestion
// Copyright (C) 2023 - 2024 by the deal.II authors
```
(Here and elsewhere.)

File: tests/fe/fe_nedelec_sz_hanging_nodes_on_inhomogene_boundary.cc [hunk 27]
```diff
@@ -0,0 +1,414 @@
+// ---------------------------------------------------------------------
```
Comment (kronbichler) [inline tests/fe/fe_nedelec_sz_hanging_nodes_on_inhomogene_boundary.cc:1]: Would you mind renaming this file into `fe_nedelec_sz_hanging_nodes_on_inhomogeneous_boundary.cc`?

**Review by kronbichler — COMMENTED** [review]
I think this looks good. This was a massive amount of work to get right, so :+1: for that! I looked into the function that computes the HN constraints and it does a lot of repetitive things, both compared to the `making_oldstyle_hanging_node_constraints` and for the various cases within the new code, but then there are several differences at some places, so I would not be able to point out another variant that could re-use some of this code. Thus, I am fine with getting this into the library in the present form, apart from the smaller adjustments outlined below. I will have another look once you've addressed them.

File: doc/news/changes/minor/20231128SebastianKinnewig [hunk 28]
```diff
@@ -0,0 +1,3 @@
+New: Support for hanging nodes in FE_NedelecSZ.
+<br>
```
[the remark this answers is not in the document]
Comment (kinnewig) [reply doc/news/changes/minor/20231128SebastianKinnewig:2]: Thank you :)

I moved that to the `major` category.

File: source/dofs/dof_tools_constraints.cc [hunk 29]
```diff
@@ -600,6 +600,16 @@ namespace DoFTools
...
 
 
+    template <typename number>
+    void
+    make_hanging_node_constraints_nedelec(const dealii::DoFHandler<1> &,
+                                          AffineConstraints<number> &,
+                                          std::integral_constant<int, 1>)
+    {
+      // nothing to do for regular dof handlers in 1d
+    }
```
[the remark this answers is not in the document]
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc:610]: I run a quick test. They are indeed unnecessary. So, I removed them.

File: source/dofs/dof_tools_constraints.cc [hunk 30]
```diff
@@ -1070,6 +1101,1290 @@ namespace DoFTools
...
+        depends_on;
+
+      // loop over all lines; only on lines there can be constraints. We do so
+      // by looping over all active cells and checking whether any of the faces
+      // are refined which can only be from the neighboring cell because this
+      // one is active. In that case, the face is subject to constraints
+      //
+      // note that even though we may visit a face twice if the neighboring
+      // cells are equally refined, we can only visit each face with hanging
+      // nodes once
```
[the remark this answers is not in the document]
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc:1124]: Yes, the functions are based on `make_oldstyle_hanging_node_constraints`.

I added the following remark (both to the 2D and 3D case):
```
Parts of this function are very similar to 
make_oldstyle_hanging_node_constraints.
Therefore, only the parts that differ from the 
make_oldstyle_hanging_node_constraints are commented on. 
```

And I removed (or at least shortened) the comments that are copied from `make_oldstyle_hanging_node_constraints`

File: source/dofs/dof_tools_constraints.cc (hunk 21 above)
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc:1501]: There is no real reason for that difference. Therefore, I unified it.
The direction is now always indicated by a `bool`.

File: source/dofs/dof_tools_constraints.cc (hunk 22 above)
Comment (kinnewig) [reply source/dofs/dof_tools_constraints.cc:1565]: Okay, I moved all comments into separate lines.

File: source/fe/fe_nedelec_sz.cc (hunk 24 above)
Comment (kinnewig) [reply source/fe/fe_nedelec_sz.cc:54]: I am not sure anymore why I placed that `TODO`. I guess I was not sure if I could keep using that assumption.

I removed that TODO.

File: tests/fe/fe_nedelec_sz_hanging_nodes_convergence.cc (hunk 26 above)
Comment (kronbichler) [reply tests/fe/fe_nedelec_sz_hanging_nodes_convergence.cc]: The end year is not so important, because we have scripts doing it automatically. We try to be reasonably precise for the first year a file was created (which is why I made the comment), so feel free to make. But it is not super important, so either way is fine.

**Comment by agrayver** [comment]
btw, I forgot there is the `GridGenerator::non_standard_orientation_mesh()` which might have been another test here. Could be added later.

**Comment by kronbichler** [comment]
Let us merge this PR now; I think this is a great progress, possible follow-up topics can be addressed later.
